r2-core/src/main/java/com/linkedin/r2/util/ServerRetryTracker.java

```diff
@@ -0,0 +1,158 @@
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.r2.util;
+
+import com.linkedin.util.clock.Clock;
+import java.util.LinkedList;
+import org.checkerframework.checker.lock.qual.GuardedBy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Stores the number of requests categorized by number of retry attempts. It uses the information to estimate
+ * a ratio of how many requests are being retried in the cluster. The ratio is then compared with
+ * {@link ServerRetryTracker#_maxRequestRetryRatio} to make a decision on whether or not to retry in the
+ * next interval. When calculating the ratio, it looks at the last {@link ServerRetryTracker#_aggregatedIntervalNum}
+ * intervals by aggregating the recorded requests.
+ */
+public class ServerRetryTracker
+{
+  private static final Logger LOG = LoggerFactory.getLogger(ServerRetryTracker.class);
+  private final int _retryLimit;
+  private final int _aggregatedIntervalNum;
+  private final double _maxRequestRetryRatio;
+  private final long _updateIntervalMs;
+  private final Clock _clock;
+
+  private final Object _counterLock = new Object();
+  private final Object _updateLock = new Object();
+
+  @GuardedBy("_updateLock")
+  private volatile long _lastRollOverTime;
+  private boolean _isBelowRetryRatio;
+
+  @GuardedBy("_counterLock")
+  private final LinkedList<int[]> _retryAttemptsCounter;
+  private final int[] _aggregatedRetryAttemptsCounter;
+
+  public ServerRetryTracker(int retryLimit, int aggregatedIntervalNum, double maxRequestRetryRatio, long updateIntervalMs, Clock clock)
+  {
+    _retryLimit = retryLimit;
+    _aggregatedIntervalNum = aggregatedIntervalNum;
+    _maxRequestRetryRatio = maxRequestRetryRatio;
+    _updateIntervalMs = updateIntervalMs;
+    _clock = clock;
+
+    _lastRollOverTime = clock.currentTimeMillis();
+    _isBelowRetryRatio = true;
+
+    _aggregatedRetryAttemptsCounter = new int[_retryLimit + 1];
+    _retryAttemptsCounter = new LinkedList<>();
+    _retryAttemptsCounter.add(new int[_retryLimit + 1]);
+  }
+
+  public void add(int numberOfRetryAttempts)
+  {
+    if (numberOfRetryAttempts <= _retryLimit)
+    {
+      synchronized (_counterLock)
+      {
+        _retryAttemptsCounter.getLast()[numberOfRetryAttempts] += 1;
+      }
+    } else
+    {
+      LOG.warn("Unexpected number of retry attempts: " + numberOfRetryAttempts + ", current retry limit: " + _retryLimit);
+      _retryAttemptsCounter.getLast()[_retryLimit] += 1;
```

Comment:
Wondering why isn't this line under the counterlock?

Reply:
Good catch. Fixed